src/nvim/eval/encode.c

```diff
@@ -0,0 +1,1291 @@
+#include "nvim/ascii.h"
+#include "nvim/vim.h"  // For _()
+#include "nvim/lib/kvec.h"
+
+#define ga_concat(a, b) ga_concat(a, (char_u *)b)
+#define utf_ptr2char(b) utf_ptr2char((char_u *)b)
+#define utf_ptr2len(b) ((size_t)utf_ptr2len((char_u *)b))
+#define utf_char2len(b) ((size_t)utf_char2len(b))
+#define string_convert(a, b, c) \
+      ((char *)string_convert((vimconv_T *)a, (char_u *)b, c))
+#define convert_setup(vcp, from, to) \
+    (convert_setup(vcp, (char_u *)from, (char_u *)to))
+
+/// Structure representing current VimL to messagepack conversion state
+typedef struct {
+  enum {
+    kMPConvDict,   ///< Convert dict_T *dictionary.
+    kMPConvList,   ///< Convert list_T *list.
+    kMPConvPairs,  ///< Convert mapping represented as a list_T* of pairs.
+  } type;
+  union {
+    struct {
+      dict_T *dict;    ///< Currently converted dictionary.
+      hashitem_T *hi;  ///< Currently converted dictionary item.
+      size_t todo;     ///< Amount of items left to process.
+    } d;  ///< State of dictionary conversion.
+    struct {
+      list_T *list;    ///< Currently converted list.
+      listitem_T *li;  ///< Currently converted list item.
+    } l;  ///< State of list or generic mapping conversion.
+  } data;  ///< Data to convert.
+} MPConvStackVal;
+
+/// Stack used to convert VimL values to messagepack.
+typedef kvec_t(MPConvStackVal) MPConvStack;
+
+const char *const encode_special_var_names[] = {
+  [kSpecialVarNull] = "null",
+  [kSpecialVarTrue] = "true",
+  [kSpecialVarFalse] = "false",
+};
+
+#ifdef INCLUDE_GENERATED_DECLARATIONS
+# include "eval/encode.c.generated.h"
+#endif
+
+/// Msgpack callback for writing to readfile()-style list
+int encode_list_write(void *data, const char *buf, size_t len)
+{
+  if (len == 0) {
+    return 0;
+  }
+  list_T *const list = (list_T *) data;
+  const char *const end = buf + len;
+  const char *line_end = buf;
+  listitem_T *li = list->lv_last;
+  do {
+    const char *line_start = line_end;
+    line_end = xmemscan(line_start, NL, (size_t) (end - line_start));
+    char *str = NULL;
+    if (line_end != line_start) {
+      const size_t line_length = (size_t) (line_end - line_start);
+      if (li == NULL) {
+        str = xmemdupz(line_start, line_length);
+      } else {
+        const size_t li_len = (li->li_tv.vval.v_string == NULL
+                               ? 0
+                               : STRLEN(li->li_tv.vval.v_string));
+        li->li_tv.vval.v_string = xrealloc(li->li_tv.vval.v_string,
+                                           li_len + line_length + 1);
+        str = (char *) li->li_tv.vval.v_string + li_len;
+        memcpy(str, line_start, line_length);
+        str[line_length] = 0;
+      }
+      memchrsub(str, NUL, NL, line_length);
+    }
+    if (li == NULL) {
+      list_append_allocated_string(list, str);
+    } else {
+      li = NULL;
+    }
+    if (line_end == end - 1) {
+      list_append_allocated_string(list, NULL);
+    }
```

Comment:
This block could be moved out of the loop. 

We also could simplify the loop, if we continue the initial last list element before the loop. All other list elements do not need to be continued.

```diff
diff --git a/src/nvim/eval/encode.c b/src/nvim/eval/encode.c
index 0bde656..5f898ec 100644
--- a/src/nvim/eval/encode.c
+++ b/src/nvim/eval/encode.c
@@ -75,36 +75,38 @@ int encode_list_write(void *data, const char *buf, size_t len)
   const char *const end = buf + len;
   const char *line_end = buf;
   listitem_T *li = list->lv_last;
-  do {
+
+  // Continue the last list element
+  if (li != NULL) {
+    line_end = xmemscan(buf, NL, len);
+    if (line_end != buf) {
+      const size_t line_length = (size_t)(line_end - buf);
+      char *str = (char *)li->li_tv.vval.v_string;
+      const size_t li_len = (str == NULL ? 0 : strlen(str));
+      li->li_tv.vval.v_string = xrealloc(str, li_len + line_length + 1);
+      str = (char *)li->li_tv.vval.v_string + li_len;
+      memcpy(str, buf, line_length);
+      str[line_length] = 0;
+      memchrsub(str, NUL, NL, line_length);
+    }
+    line_end++;
+  }
+
+  while (line_end < end) {
     const char *line_start = line_end;
     line_end = xmemscan(line_start, NL, (size_t) (end - line_start));
     char *str = NULL;
     if (line_end != line_start) {
-      const size_t line_length = (size_t) (line_end - line_start);
-      if (li == NULL) {
-        str = xmemdupz(line_start, line_length);
-      } else {
-        const size_t li_len = (li->li_tv.vval.v_string == NULL
-                               ? 0
-                               : STRLEN(li->li_tv.vval.v_string));
-        li->li_tv.vval.v_string = xrealloc(li->li_tv.vval.v_string,
-                                           li_len + line_length + 1);
-        str = (char *) li->li_tv.vval.v_string + li_len;
-        memcpy(str, line_start, line_length);
-        str[line_length] = 0;
-      }
+      const size_t line_length = (size_t)(line_end - line_start);
+      str = xmemdupz(line_start, line_length);
       memchrsub(str, NUL, NL, line_length);
     }
-    if (li == NULL) {
-      list_append_allocated_string(list, str);
-    } else {
-      li = NULL;
-    }
-    if (line_end == end - 1) {
-      list_append_allocated_string(list, NULL);
-    }
+    list_append_allocated_string(list, str);
     line_end++;
-  } while (line_end < end);
+  }
+  if (line_end == end) {
+    list_append_allocated_string(list, NULL);
+  }
   return 0;
 }

```